### DISCUSSION TITLE
Is there a reason why Aws::Credentials is not a Aws::CredentialProvider?

### DISCUSSION THREAD

**Discussion by cervantn** [aws/aws-sdk-ruby#2710] [Q&A]
Is there any reason why [`Aws::Credentials`](https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-core/lib/aws-sdk-core/credentials.rb) does not `include` [`Aws::CredentialProvider`](https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-core/lib/aws-sdk-core/credential_provider.rb), like other classes in the SDK (e.g. [`AssumeRoleCredentials`](https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-core/lib/aws-sdk-core/assume_role_credentials.rb)) which implement the credential provider interface (implementing `#credentials` and `set?`)?

**Comment by alextwoods** [comment]
Its largely legacy - the credential providers use `.credentials` to return [Aws::Credentials](https://github.com/aws/aws-sdk-ruby/blob/version-3/gems/aws-sdk-core/lib/aws-sdk-core/credentials.rb), so we don't want those credentials to also be a provider.  However, we should have created a separate StaticCredentialProvider instead of adding a `.credentials`  method to Aws::Credentials.    Additionally, the `set?` method in `Aws::CredentialProvider` relies on calling the `set?` method of the returned `credentials`.  

Do you have a use case for needing Aws::Credentials to include Aws::CredentialsProvider?

**Reply by cervantn** [reply]
Not any sort of critical use case. I was writing some unit tests recently for a utility which returns either an `Aws::Credentials` or `Aws::CredentialProvider` instance depending on its input (it's not really important how or why). But I was hoping that I could write a single RSpec assertion for all examples like:

```ruby
it { is_expected.to be_an(Aws::CredentialsProvider) }
```

but this fails for the cases where an `Aws::Credentials` instance is returned instead. I was able to work around it with a custom matcher:

```ruby
define :return_aws_sdk_credentials do
  match do |actual|
    actual.is_a?(Aws::CredentialProvider) || actual.is_a?(Aws::Credentials)
  end
end

...

it { is_expected.to return_aws_sdk_credentials }
```

but honestly I wish I didn't need to think about these cases differently. In all of the code which uses this utility, I don't need to think about it because I can call `#credentials` in either case. I also recognize that I could instead be doing a `respond_to` assertion, but I guess I'm just not a big fan of duck in unit tests.
